lib/WWW/Mechanize.pm

```diff
@@ -1454,6 +1433,59 @@ sub form_with_fields {
     }
 }
 
+
+=head2 $mech->form_with( $attr1 => $value1, $attr2 => $value2, ... )
+
+Searches for forms with arbitrary attribute/value pairs within the E<lt>formE<gt>
+tag.
+(Currently does not work for attribute C<action> due to implementation details
+of L<HTML::Form>.)
+When given more than one pair, all criteria must match.
+Using C<undef> as value means that the attribute in question may not be present.
+
+If it is found, the form is returned as an L<HTML::Form> object and set internally
+for later used with Mech's form methods such as C<L</field()>> and C<L</click()>>.
+
+Returns undef if no form is found.
+
+
+=cut
+
+sub form_with {
+    my ( $self, %spec ) = @_;
+
+    my @forms = $self->forms or return;
+    while ( my ( $attr, $expected_value ) = each %spec ) {
+        @forms = grep {
+            my $actual_value = $_->attr($attr);
+            if ( defined $expected_value ) {
+                defined $actual_value && $actual_value eq $expected_value;
+            }
+            else {
+                !defined $actual_value;
+            }
+        } @forms or return;
```

Comment:
Thanks for your suggestion, @karenetheridge!  However, this would IMHO impose two problems in case an attribute searched for is not present in the page:
* "Use of uninitialized value in string eq" warnings would be emitted.
* The case "attribute is present with empty value" would not be distinguished from "attribute is not present".

I have reworked this `grep` in the meantime, though:
https://github.com/libwww-perl/WWW-Mechanize/pull/48/commits/db3e94d96b4e4223d1b3359a326ed1713867224d
Maybe you like the new version.